compiler/rustc_const_eval/src/interpret/intern.rs

```diff
@@ -168,10 +168,22 @@ impl<'rt, 'mir, 'tcx: 'mir, M: CompileTimeMachine<'mir, 'tcx, const_eval::Memory
         mplace: &MPlaceTy<'tcx>,
         fields: impl Iterator<Item = InterpResult<'tcx, Self::V>>,
     ) -> InterpResult<'tcx> {
-        // ZSTs cannot contain pointers, so we can skip them.
-        if mplace.layout.is_zst() {
+        // We want to walk the aggregate to look for reference types to intern. While doing that we
+        // also need to take special care of interior mutability.
+        //
+        // As an optimization, however, if the allocation does not contain any references: we don't
+        // need to do the walk. It can be costly for big arrays for example (e.g. issue #93215).
+
+        let Some((size, align)) = self.ecx.size_and_align_of_mplace(&mplace)? else {
```

Comment:
FWIW, this call here might be not entirely cheap. So doing it recursively on all aggregates we encounter might be a bad idea. Maybe leave a FIXME?

An alternative might be to only go check this if the type is `Array` or `Slice`, as those are usually the large types.

Reply:
I've implemented this narrowing in https://github.com/rust-lang/rust/pull/97585/commits/84bc03f1a1316abd1ae3ee0b345c93b11c6cd579.